### PR TITLE
Renaming of artifacts on upload and download

### PR DESCRIPTION
**Pull request by toadzky**
Addresses #25. 

## Uploading

```yml
 steps:
   - plugins:
       artifacts#v1.2.0:
         upload: log/foo.log -> foo.log
 ```

## Downloading:

```yml
 steps:
   - plugins:
       artifacts#v1.2.0:
         download: foo.log -> /tmp/foo.log
 ```

### CODE REVIEW

File: hooks/post-command [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/bin/bash -x
```
Comment (lox) [inline hooks/post-command]: Can we ditch the `-x`?

**Comment by lox** [comment]
Thanks for this @toadzky! Would it be possible to add to the README the new syntax? It's hard to extract from the code and will make review easier!

File: hooks/post-command (hunk 1 above)
Comment (toadzky) [reply hooks/post-command]: sure

**Comment by toadzky** [comment]
Readme updated. :+1:

**Comment by lox** [comment]
I took the liberty of updating the PR description with the proposed syntax to make it easy for folks to review!

**Comment by lox** [comment]
I'm a bit nervous about using spaces and `->` to denote `from` and `to`.  What if we used maps instead? 

```
 steps:
   - plugins:
       artifacts#v1.2.0:
         upload: 
           "log/foo.log": "foo.log"
```

**Comment by keithbranton** [comment]
@lox How could that be supported without sacrificing backward compatibility.

This PR does not require spaces around the arrow, although it does ignore any that are present.

This PR, as written, will happily coexist with all your other examples.

**Comment by toolmantim** [comment]
Another alternative to this would be a `strip-directories` arguments (similar to https://github.com/buildkite/agent/issues/752), and a `to`, which would also allow for globs:

```yaml
 steps:
  # Creates dist/some-app
  - command: make
    plugins:
      - artifacts#v1.2.0:
          upload: dist/*

  - wait

  - commands:
      # Requires app/some-app for the build
      - "docker build -t some-app-$BUILDKITE_BUILDER_NUMBER ."
      - "docker push some-app-$BUILDKITE_BUILDER_NUMBER"
    plugins:
      - artifacts#v1.2.0:
          download: dist/some-app
          to: app
          strip-directories: true
```

This would also work with the standard command artifact upload syntax, because you don't need to rely on the rename in the upload:

```yaml
 steps:
  - command: make
    artifact_paths: log/app.log

  - wait

  - plugins:
      - artifacts#v1.2.0:
          download: logs/*
          to: app-logs
          strip-directories: true
```

but there's no reason it couldn't work on the upload too:

```yaml
 steps:
  - command: make
    plugins:
      - artifacts#v1.2.0:
          upload: logs/*
          to: app-logs
          strip-directories: true   

  - wait

  - plugins:
      - artifacts#v1.2.0:
          download: app-logs/*
```

**Comment by lox** [comment]
I'm inclined to add `--strip-directories` to the `buildkite-agent artifact upload` command and then expose that option in this plugin.

**Comment by toadzky** [comment]
that doesn't solve the use case though. first, that does nothing for downloads and second, it's not always about just removing the directories. it's about putting it somewhere specific, both on upload and download. it fits with the existing usage nicely, which means you can move somethings but not others (as opposed to alternatives where you have to put sub-objects). your examples are always around uploading whole folders, but that's not a use-case we are using. we have multiple files from different folders that we want to upload to different places, and we download multiple files to different places.

**Comment by toolmantim** [comment]
What if we were to cut a v2.0.0 of the plugin, not worry about backward compatibility, and change the syntax entirely to something like what @lox suggested?

For most plugins we're trying to iterate towards ideal APIs without doing lots of backwards compatibility support… especially in the early version numbers.

```yml
 steps:
   - plugins:
       artifacts#v2.0.0:
         upload: 
           "log/foo.log": "foo.log"
```

I wonder how something like this should work?

```yml
 steps:
   - plugins:
       artifacts#v2.0.0:
         upload: 
           "log/foo.log": "dir/foo.log"
```

How should it handle the `dir` directory which doesn't exist in the local checkout? If you wanted to do it from a build script you'd need to do:

```shell
ln -sf log dir
buildkite-agent artifact upload dir/foo.log
```

And if it was nested even more:

```yml
 steps:
   - plugins:
       artifacts#v2.0.0:
         upload: 
           "log/foo.log": "dir1/dir2/foo.log"
```

It'd need to do an `mkdir -p` before the symlinking:

```shell
mkdir -p dir1
ln -sf ../log dir1/dir2
buildkite-agent artifact upload dir1/dir2/foo.log
```

To support globs, if we followed the same [buildkite-agent artifact](https://buildkite.com/docs/agent/v3/cli-artifact) syntax, my previous example would actually be:

```yml
 steps:
  # Creates dist/some-app
  - command: make
    plugins:
      - artifacts#v2.0.0:
          upload:
            "dist/some-app/**/*": "some-app/"
```

Which to do in a script would be (notice we had to add the extra `**/*` to the upload ourselves)

```shell
ln -sf dist/some-app some-app 
buildkite-agent artifact upload "some-app/**/*"
```

And the second part:

```yml
  - commands:
      # Requires app for the build
      - "docker build -t some-app-$BUILDKITE_BUILDER_NUMBER ."
      - "docker push some-app-$BUILDKITE_BUILDER_NUMBER"
    plugins:
      - artifacts#v2.0.0:
          download:
            "some-app/*": "app/"
```

You think would be:

```shell
buildkite-agent artifact download "some-app/*" app/
```

but it would actually create `app/some-app/files` — we'd need a `--strip-prefx some-app/` or similar argument to the `artifact download` command:

```shell
buildkite-agent artifact download --strip-prefix "some-app/" "some-app/*" app/
```

Globs and nested directories make everything a bit trickier! We could just throw an error on globs or nested target directories? But it's a shame not to try to figure out a syntax that could work for those cases, if we can!

**Comment by toolmantim** [comment]
Thanks for your patience @toadzky and @keithbranton while we figure it out!

**Comment by toolmantim** [comment]
All the `mkdir` and `ln` commands make me think we’re missing a few arguments/primitives at the `buildkite-agent artifact` command level 🤔

**Comment by toadzky** [comment]
personally (and probably obviously as well) i prefer the `->` because i think it's easier to read and use.

however, if we did throw out backward compatibility and make the syntax more explicit, considering ease of use, if you didn't want to relocate the files, how would you do it? making users put in the same path as key and value seems cumbersome. you would need to split files by relocation or not. something like 

```yaml
  - commands:
      # Requires app for the build
      - "docker build -t some-app-$BUILDKITE_BUILDER_NUMBER ."
      - "docker push some-app-$BUILDKITE_BUILDER_NUMBER"
    plugins:
      - artifacts#v2.0.0:
          download:
            move: 
              "some-app/*": "app/"
            leave:
              - somewhere/something
```

That's not better, but it's the only thing I can think of because yaml doesn't allow mixing of list items and object key/value pairs. I really do think that the current syntax in the PR is the best way to allow users the flexibility some of us want while not impacting the ease of use for the rest of the users.

**Comment by toolmantim** [comment]
I'd really like to figure out what to do on this one!

We've another suggestion from @matthewd, which keeps the array syntax we currently use, but adds support for maps in the array, which we could limit to single file relocation (globs not supported) as first cut, just like you've done in this PR @toadzky:

Uploading:

```yml
 steps:
   - plugins:
       artifacts#v1.2.0:
         upload:
           - "log/foo.log": "foo.log"
           - "debug/*.error"
```

Downloading:

```yml
 steps:
   - plugins:
       artifacts#v1.2.0:
         download:
           - "foo.log": "/tmp/foo.log"
           - "debug/*"
```

**Comment by toolmantim** [comment]
Whoops, accidental close sorry!

**Comment by toadzky** [comment]
That's probably fine. I personally like the arrow syntax because I think it communicates the intent better, but that's not a big deal. Would you like me to update this PR to support that syntax? It should work the same as the arrow syntax, just with the `:` instead. 

Actually, if you don't object, I'm pretty sure we could support both syntaxes.

**Comment by toolmantim** [comment]
I think using an actual map, rather than string munging, would be the ideal syntax we'd prefer (to avoid string parsing and keep the config structured with YAML primitives). It also sets us up for getting the same functionality rolled into the standard BK pipeline.yml and agent.

But, I just did some testing on the suggested syntax by @matthewd, before you went and spent any time, and I think there's a problem that means we can't use that syntax at the moment.

Given the following syntax:

```yml
 steps:
   - plugins:
       artifacts#v1.2.0:
         upload:
           - "log/foo.log": "foo.log"
           - "debug/*.error"
  - wait
   - plugins:
       artifacts#v1.2.0:
         download:
           - "foo.log": "/tmp/foo.log"
           - "debug/*"
```

The agent passes through the config to the plugin, as the following environment variables:

```
BUILDKITE_PLUGIN_ARTIFACTS_UPLOAD_0_LOG/FOO.LOG=foo.log
BUILDKITE_PLUGIN_ARTIFACTS_UPLOAD_1=debug/*.error
```

```
BUILDKITE_PLUGIN_ARTIFACTS_DOWNLOAD_0_FOO.LOG=/tmp/foo.log
BUILDKITE_PLUGIN_ARTIFACTS_DOWNLOAD_1=debug/*
```

We've basically lost the case sensitivity via env var names, and still have to do string munging to extract them out.

If we wanted the original config in JSON, it's available inside `BUILDKITE_PLUGINS` but there's no way to tell which part of the JSON is for each particular plugin invocation. And we'd need to require `jq` to be installed (maybe not so bad?).

If the agent passed through a JSON of the config for each plugin invocation (`BUILDKITE_PLUGIN` perhaps?) then it'd be workable, with jq installed.

Tricky!

**Comment by toolmantim** [comment]
I think the only way forward would be have named keys, ala:

```yml
steps:
  - plugins:
      artifacts#v1.2.0:
        upload:
          - { from: "log/foo.log", to: "foo.log" }
          - "debug/*.error"
  - wait
  - plugins:
      artifacts#v1.2.0:
        download:
          - { from: "foo.log", to: "/tmp/foo.log" }
          - "debug/*"
```

And we can probably leave "blob/folder support" as a todo.

We'd prefer to leave out `->` string support, and aim for a more structured config based on the YAML primitives.

File: README.md [hunk 2]
```diff
@@ -60,6 +104,11 @@ The job UUID or name to download the artifact from.
 
 The build UUID to download the artifact from.
 
+### Relocation
+
+If a file needs to be renamed or moved before upload or after download, a special pattern is supported: `source -> destination`.
```
Comment (lox) [inline README.md]: Sounds like this might not be true anymore?

Comment (toadzky) [reply README.md]: good catch. i'll fix the readme

**Comment by lox** [comment]
Thanks for your patience @toadzky!

**Comment by toolmantim** [comment]
This will be an exciting release! ⚡️

**Comment by toadzky** [comment]
:+1: any idea when it will be released?

**Comment by toolmantim** [comment]
I've updated the instructions for running that locally in 30178dc and made some readme fixups for the plugin linter, but I think the plugin.yml is out of date now?

**Comment by toadzky** [comment]
what's wrong with the plugin.yml and how would you like it to be changed?

**Comment by toolmantim** [comment]
@toadzky if you run `bk run` you'll see an error that the plugin linter has caught, that the plugin.yml `properties` JSON schema no longer matches the examples in the readme.

We'll need to fix the schema so it that matches all the variations we've added (I just clarified the formats again in 309df16e492a2f2d5774eea5325cab1c876cac90 — based on the readme examples)

**Pull request merged**